rfcs/0000_monty_versioning.md

```diff
@@ -0,0 +1,120 @@
+- Start Date: 2025-02-20
+- RFC PR: (leave this empty, it will be filled in after RFC is merged)
+
+# Summary
+
+Begin versioning Monty using [semantic versioning](https://semver.org/).
+
+# Motivation
+
+We want papers published by the Thousand Brains Project to be reproducible. The likely approach the project will adopt is to create a repository for each paper and include the code to reproduce the paper in the repository. When the paper relies on `tbp.monty`, it should be clear which version of `tbp.monty` is needed to reproduce the paper. Ideally, the paper repository declares a specific `tbp.monty` version as a dependency that is installed like any other dependency.
+
+# Explanation
+
+Monty uses [semantic versioning](https://semver.org/) for versions.
+
+The intent behind semantic versioning is to provide a way to rapidly and tersely communicate to people and machines the compatibility of a software change.
+
+In summary, the version number is `MAJOR.MINOR.PATCH`.
+
+* `MAJOR` is incremented when backwards-incompatible (breaking) changes are made.
+* `MINOR` is incremented when new functionality is added in a backwards-compatible manner.
+* `PATCH` is incremented when backwards-compatible bug fixes are made.
+* CI, tools, infrastructure, documentation, or RFC changes do not increment the version number.
+
+The current version can be found in [`src/tbp/monty/__init__.py`](https://github.com/thousandbrainsproject/tbp.monty/blob/main/src/tbp/monty/__init__.py) as the `__version__` variable.
+
+## Updating the version
+
+### Constraints
+
+Here are some constraints related to updating the version:
+
+* We want the current version to be in the source code.
+* The `main` branch is protected and changes can be pushed only via Pull Requests.
+
+### Process
+
+1. Create a Pull Request that ***only*** updates the `__version__` variable in [`src/tbp/monty/__init__.py`](https://github.com/thousandbrainsproject/tbp.monty/blob/main/src/tbp/monty/__init__.py).
```

Comment:
Can you elaborate why we would want a PR with only that change instead if including the version number update in the PR with the change that triggers the version update? If we have a PR with a breaking change + a PR with the version number update, which PR would be merged first? I would assume we don't want to add the breaking change into main under the old version. So then we would increment the version before introducing the change? Why couldn't we increment the version in the same PR as the change and have the discussion on whether the change warrants a version update there?

Reply:
I also added this future possibility in https://github.com/thousandbrainsproject/tbp.monty/pull/186/commits/7ff85b63a3873e5294e5be74bcdeb2792cdfc609

In the future, integrating into the `main` branch may become cumbersome if the pace of commits to `main` is high. In that case, we will update our versioning strategy. One way to do this is to create a `release` branch that is used for releasing new versions. This note is intended to highlight the possibility, but if we get there, we will get into the details then.